src/main/c/NativePtr.cpp

```diff
@@ -0,0 +1,367 @@
+#include "hail/NativePtr.h"
+#include "hail/NativeObj.h"
+#include <cassert>
+#include <cstdio>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+#include <jni.h>
+
+namespace hail {
+
+// JNI interface functions corresponding to Scala NativeBase and NativePtr
+
+#define NATIVEFUNC(typ) \
+  extern "C" JNIEXPORT typ JNICALL
+
+namespace {
+
+void check_assumptions();
+
+class NativePtrInfo {
+public:
+  jclass class_ref_;
+  jfieldID addrA_id_;
+  jfieldID addrB_id_;
+  
+public:
+  NativePtrInfo(JNIEnv* env, int line) {
+    auto cl = env->FindClass("is/hail/nativecode/NativeBase");
+    class_ref_ = (jclass)env->NewGlobalRef(cl);
+    env->DeleteLocalRef(cl);
+    addrA_id_ = env->GetFieldID(class_ref_, "addrA", "J");
+    addrB_id_ = env->GetFieldID(class_ref_, "addrB", "J");
+    check_assumptions();
+  }
+};
+
+// WARNING: I observe that when this gets loaded as a shared library on Linux,
+// we see two distinct NativePtrInfo objects at different addresses.  That is
+// extremely weird, but by putting all the initialization into the constructor
+// we make sure that both get correctly initialized.  But that behavior might
+// cause trouble in other code, so we need to watch out for it.
```

Comment:
This is quite disturbing.  Do you still not know why this is happening?  The only thing I can think of is that the .so is getting loaded twice (by two difference class loaders, maybe?)  Is that possible?